### PR TITLE
[#45] Add nettest scenario that covers all entryoints

### LINKED ISSUE #45: Extend nettest scenario to cover all entrypoints
# Clarification and motivation

We have a very simple nettest scenario that only calls `transfer`. That's not sufficient, we should at least call all entrypoints to know they don't exceed gas consumption.

# Acceptance criteria

Our nettest-based tests call all entrypoints present in the contract.

Please don't start before #37, #35, #31.

### PR DESCRIPTION
**Pull request by awkure**
Problem: We need to test our stablecoin contract against
real network to ensure that all the entryopints are called
properly.

Solution: Add nettest scenario that calls all entryopints.

## Related issue(s)

Resolves #45 

#### Related changes (conditional)

- Tests
  - [X] If I added new functionality, I added tests covering it.
  - [X] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

- Documentation
  - [X] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)
    - Haddock
  - [X] I updated [changelog](../tree/master/ChangeLog.md) unless I am sure my changes are
        not essential.

#### Stylistic guide (mandatory)

- [X] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Where-and-how-to-commit-your-work-58f8973a4b3142c8abbd2e6fd5b3a08e).
- [X] My code complies with the [style guide](../tree/master/docs/code-style.md).

### CODE REVIEW

**Comment by gromakovsky** [comment]
Tests are failing because `newAddress` does not send enough XTZ to cover transaction costs. The primary reason is that we use too high transaction costs. It's fixed [here](https://gitlab.com/morley-framework/morley/-/merge_requests/443). I suppose we can wait for that MR to be merged, hopefully it will resolve the problem. If it doesn't (if we make too many calls from some address) you can also manually transfer more from `nettest` to addresses that need more.

File: haskell/nettest/Nettest.hs [hunk 1]
```diff
@@ -54,6 +95,160 @@ scTransferScenario constructInitialStorage contract_ = uncapsNettest $ do
...
+
+    setSafelist :: Address -> Address -> capsM ()
+    setSafelist from safelistAddress = callFrom (AddressResolved from) sc (ep "set_safelist") (Just safelistAddress)
+
+    unsetSafelist :: Address -> capsM ()
+    unsetSafelist from = callFrom (AddressResolved from) sc (ep "set_safelist") (Nothing :: Maybe Address)
+
+  let
+    transferScenario = do
+      comment "Transfer from superuser to superuser "
```
Comment (gromakovsky) [inline haskell/nettest/Nettest.hs:147]: IMO it looks a bit confusing when you print `"Transfer from superuser to superuser"`, make a transfer from superuser to superuser and then make other transfers before printing something else. If `callTransfer superuser owner1 40` fails, one will think that "Transfer from superuser to superuser failed", but in reality a different transfer failed. So IMO if there is a comment `X` then all actions between `X` and the next comment `Y` (if there is `Y`) should somehow relate to `X`.

File: haskell/nettest/Nettest.hs [hunk 2]
```diff
@@ -54,6 +95,160 @@ scTransferScenario constructInitialStorage contract_ = uncapsNettest $ do
...
+      callTransfer superuser owner1 40
+
+      callTransfers
+        [ (#from_ .! owner1, [(#to_ .! owner2, #amount .! 10)])
+        , (#from_ .! owner2, [(#to_ .! owner3, #amount .! 10)])
+        , (#from_ .! owner3, [(#to_ .! owner1, #amount .! 10)])
+        ]
+
+      comment "Pausing contract for transfers"
+      expectFailed $ pause owner1 -- Not enough permissions
```
Comment (gromakovsky) [inline haskell/nettest/Nettest.hs:158]: One day we'll be able to specify the exact expected error :coolstory:

File: haskell/src/Indigo/Contracts/Safelist.hs [hunk 3]
```diff
@@ -16,6 +16,7 @@ import qualified Data.Set as Set
 
 import Indigo as Indigo
 import Lorentz.Run (Contract)
+import Prelude ()
```
Comment (gromakovsky) [inline haskell/src/Indigo/Contracts/Safelist.hs:19]: Erm, did it become necessary?

File: haskell/nettest/Nettest.hs (hunk 1 above)
Comment (gromakovsky) [reply haskell/nettest/Nettest.hs:156]: Can you change it to "Transfers from superuser"? Because there are 2 transfers in this group, both are from `superuser`, but only one is to `superuser`.

**Comment by awkure** [comment]
Note, that I've used a minor [hotfix](https://gitlab.com/morley-framework/morley/-/commit/ec0a1f2d78108c1e82d837b6e1cbd86cf41b5d7e) in morley to make updated integrational tests pass. For the discussion you can refer to our [slack thread](https://serokell.slack.com/archives/CUMAQ3PL4/p1592934064227100)

File: haskell/test-common/Lorentz/Contracts/Test/Common.hs [hunk 4]
```diff
@@ -225,7 +225,7 @@ skipTest = do
     dummyContract = L.unpair L.# L.drop L.# L.nil L.# L.pair
   c <- lOriginate (L.defaultContract dummyContract) "skip test dummy" () (unsafeMkMutez 0)
   lCallEP c CallDefault ()
-  validate (Right expectAnySuccess)
+  expectNoUpdates
```
Comment (gromakovsky) [inline haskell/test-common/Lorentz/Contracts/Test/Common.hs:228]: I think `skipTest` is just not needed anymore. It should be possible to just do nothing now, after recent `Michelson.Test` changes.

File: haskell/test/Lorentz/Contracts/Test/Management.hs [hunk 5]
```diff
@@ -756,8 +759,7 @@ managementSpec originate = do
               ]
 
           withSender wallet1 $ lCallEP stablecoinContract (Call @"Mint") mintings
-
-          validate (Right expectAnySuccess)
+          pass
```
Comment (gromakovsky) [inline haskell/test/Lorentz/Contracts/Test/Management.hs:762]: All these `pass` usages are redundant, aren't they?

File: haskell/src/Indigo/Contracts/Safelist.hs (hunk 3 above)
Comment (gromakovsky) [reply haskell/src/Indigo/Contracts/Safelist.hs:19]: @awkure ping.

File: haskell/test/Lorentz/Contracts/Test/Management.hs [hunk 6]
```diff
@@ -471,8 +469,9 @@ managementSpec originate = do
           $ addAccount (wallet1, (commonOperators, 10))
           $ defaultOriginationParams
       withOriginated originate originationParams $ \stablecoinContract -> do
-        withSender wallet1 $ lCallEP stablecoinContract (Call @"Burn") [ 10, 10 ]
-        validate . Left $ mgmInsufficientBalance
+        -- withSender wallet1 $ lCallEP stablecoinContract (Call @"Burn") [ 10, 10 ]
```
Comment (gromakovsky) [inline haskell/test/Lorentz/Contracts/Test/Management.hs:472]: Looks like a leftover.

File: haskell/src/Indigo/Contracts/Safelist.hs (hunk 3 above)
Comment (awkure) [reply haskell/src/Indigo/Contracts/Safelist.hs:19]: Yes, because it conflicts with `import Indigo` recently updated operators

Comment (gromakovsky) [reply haskell/src/Indigo/Contracts/Safelist.hs:19]: Alright, maybe it would be better to enable `NoImplicitPrelude` globally for the lib, but I am not sure, so let's not do it for now.

File: haskell/test/Lorentz/Contracts/Test/FA2.hs [hunk 7]
```diff
@@ -607,9 +586,57 @@ fa2Spec fa2Originate = do
             let isOperatorQuery = mkView (#operator .! operatorParam) consumer
             lCallEP fa2contract (Call @"Is_operator") isOperatorQuery
 
-            (validate . Right $
-              lExpectViewConsumerStorage consumer
-                [(#operator .! operatorParam, #is_operator .! True)])
+            lExpectViewConsumerStorage consumer
+                [(#operator .! operatorParam, #is_operator .! True)]
+
+  -- Check whether "update operator", "remove operator" operations are executed only by contract owner.
```
Comment (gromakovsky) [inline haskell/test/Lorentz/Contracts/Test/FA2.hs:592]: Are these tests new? Why did you add them as part of morley update?

File: haskell/nettest/Nettest.hs [hunk 8]
```diff
@@ -50,10 +92,186 @@ scTransferScenario constructInitialStorage contract_ = uncapsNettest $ do
...
+    callTransfers = mapM_ $ \(from@(arg #from_ -> from_), destinations) ->
+      callFrom
+        (AddressResolved from_)
+        sc
+        (ep "transfer")
+        (constructTransfersFromSender from destinations)
+
+    configureMinter :: Address -> Address -> Maybe Natural -> Natural -> capsM ()
+    configureMinter from for expectedAllowance newAllowance =
+      callFrom (AddressResolved from) sc (ep "configure_minter") (#minter .! for, (#current_minting_allowance .! expectedAllowance, #new_minting_allowance .! newAllowance))
```
Comment (gromakovsky) [inline haskell/nettest/Nettest.hs:112]: Some lines here are :so-big:, this one is 173 characters long, it's not ok.

Comment (awkure) [reply haskell/nettest/Nettest.hs:112]: :bruh:

**Comment by gromakovsky** [comment]
Also please check whether there are warnings. IIRC CI doesn't check that for some reason.

File: haskell/test/Lorentz/Contracts/Test/FA2.hs (hunk 7 above)
Comment (awkure) [reply haskell/test/Lorentz/Contracts/Test/FA2.hs:592]: Well, this is weir how this appeared in management tests and got merged onto master. I'll make a separate commit for it

Comment (gromakovsky) [reply haskell/test/Lorentz/Contracts/Test/FA2.hs:591]: And where is this promised separate commit? :pled:
